### PR TITLE
Fix cidr mask conversion issue for 0.0.0.0/0 and add tests #12005

### PR DESCRIPTION
**Pull request by ruflin**
This fixes the issue raised in #12005 and adds test to verify it.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java [hunk 1]
```diff
@@ -148,7 +152,7 @@ public IPv4RangeBuilder addUnboundedFrom(String from) {
         int to = from + (~mask);
         long longTo = intIpToLongIp(to) + 1; // we have to +1 here as the range is non-inclusive on the "to" side
         if (longTo == InternalIPv4Range.MAX_IP) {
-            longTo = -1;
+            longTo =  InternalIPv4Range.MAX_IP - 1;
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java]: I understand the above change, but not this one: what does it fix?

Comment (ruflin) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java]: Setting the value to `-1` leads to `null` as the to IP address. I assumed the expected IP address should be 255.255.255.255. But now that I rethink it there is a potential issue with that. As it is mentioned on line 153 the range is non inclusive. So it should be 256.0.0.0 which also correlates with `InternalIPv4Range.MAX_IP`. But I think this is actually not a valid IP address which is the reason why the system converts it to `null`. Going with the above solution would exclude the IP address 255.255.255.255.

As it is mentioned in the docs block `-1` means unbound end so you are right and the above change should not be needed. I will think of a test to verify that if the IP is 255.255.255.255 it is inside the range and will update the pull request.

As `-1` and `InternalIPv4Range.MAX_IP` are both converted to `null` I will check if the if clause is even needed.

**Comment by ruflin** [comment]
@jpountz I updated the tests. Based on https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IpRangeParser.java#L133 the -1 has to stay.

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java [hunk 2]
```diff
@@ -869,4 +870,77 @@ public void emptyAggregation() throws Exception {
...
+        assertThat((String) bucket.getKey(), equalTo("0.0.0.0/0"));
+        assertThat(bucket.getFromAsString(), nullValue());
+        assertThat(bucket.getToAsString(), nullValue());
+        assertThat(((Number) bucket.getTo()).doubleValue(), equalTo(Double.POSITIVE_INFINITY));
+        assertEquals(255l, bucket.getDocCount());
+    }
+
+
+    @Test
+    public void mask0SpecialIps() throws Exception {
```
Comment (ruflin) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java]: @jpountz This creates and index and adds documents to it. Should the first part be moved to the `setupSuiteScopeCluster(*)`?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java (hunk 1 above)
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/range/ipv4/IPv4RangeBuilder.java]: Thanks.

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java (hunk 2 above)
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java]: I missed it, indeed it should be moved to setupSuiteScopeCluster

**Comment by jpountz** [comment]
@ruflin Would you mind signing the [CLA](https://www.elastic.co/contributor-agreement). This will be required so that your pull request can be merged.

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java (hunk 2 above)
Comment (ruflin) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/IPv4RangeTests.java]: I will move it up.

**Comment by ruflin** [comment]
@jpountz I just signed it, but haven't received a confirmation yet.

**Comment by ruflin** [comment]
Test moved up and CLA signed.

**Comment by jpountz** [comment]
Thanks!

> The CLA should probably be updated to Elastic. It is still Elasticsearch.

Elastic is the trade name of the company, but Elasticsearch BV is still the registed name of the company, this is why the CLA still mentions Elasticsearch BV.

**Pull request merged**